### ISSUE TITLE
Tabby (ai coding assistant) segmentation fault

### ISSUE THREAD

**Issue by meditans** [NixOS/nixpkgs#333538]
### Describe the bug
When trying this command tabby segfaults:
```
$ tabby serve --model TabbyML/StarCoder-1B
2024-08-09T21:36:29.342833Z  INFO tabby::serve: crates/tabby/src/serve.rs:116: Starting server, this might take a few minutes...
Segmentation fault (core dumped)
```

### Steps To Reproduce
I have a flake-based system. I tried installing tabby with the following variants:
```
tabby
```
```
(tabby.override { acceleration = "cuda"; })
```
```
((tabby.override { acceleration = "cuda"; }).overrideAttrs (oldAttrs: rec {
  version = "0.15.0";
}))
```
```
(tabby.overrideAttrs (oldAttrs: rec {
  version = "0.15.0";
}))
```

All lead to segmentation fault.

### Expected behavior
I'd like to avoid the segfault. Any suggestion on debugging?

### Notify maintainers
@ghthor 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
$ nix run nixpkgs#nix-info -- -m
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.215, NixOS, 24.05 (Uakari), 24.05.20240427.58a1abd`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/801l7gvdz7yaibhjsxqx82sc7zkakjbq-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by ghthor** [comment]
Not surprised, it’s broken from updates to llama-cpp. And on 0.13 (or 0.12)
tabby changed from linking with llama-cpp to running llama-cpp as a sub
process which I haven’t incorporated into the nix package yet so this is
the likely reason for the seg faults.

Apologies about this being broken, I haven’t had the time to fix it yet;
never imagined that the build would break so often considering tabby was a
rust binary.

**Comment by wsxiaoys** [comment]
Just notice this issue - I would recommend only incorp Tabby's own rust binary building process in Nix - and reply on llama.cpp's Nix distribution directly, instead of building it as part of tabby

**Comment by meditans** [comment]
I agree with @wsxiaoys on this, also because since we already have `ollama` packaged in nix as a daemon, it's very easy to just offload the inference part to that. In fact, locally, I built tabby with the flake below, and I'm relying on ollama for inference.
```nix
{
  description = "tabbyML build";

  inputs = {
    nixpkgs.url = "nixpkgs/nixos-unstable"; # Define the Nix Packages collection
  };

  outputs =
    { self, nixpkgs }:
    let
      system = "x86_64-linux"; # Specify the system for which this flake is intended
      pkgs = import nixpkgs { inherit system; };
    in
    {
      devShells.${system}.default = pkgs.mkShell {
        buildInputs = with pkgs; [
          rustc
          cargo
          protobuf
          sqlite
          graphviz
          openssl.dev
          pkg-config
        ];
      };
    };
}
```

**Comment by ghthor** [comment]
@wsxiaoys reading up on tabby release, I see that tabby now supports an http API for the backend. I assume what you mean with your statement is to just configure the nixpkg to use that?

So questions about that;

1. just build tabby with cpu acceleration only?
1. how to control which GPU type gets used in the backend
    1. for instance I have both amd and nvidia gpus
    1. does that come down to configuration of the backend service (ollama, etc.)

**Comment by wsxiaoys** [comment]
> just build tabby with cpu acceleration only?
Yes

> how to control which GPU type gets used in the backend
> for instance I have both amd and nvidia gpus
> does that come down to configuration of the backend service (ollama, etc.)

Right - it comes to the backend service configuration instead.

**Comment by ghthor** [comment]
Ok here's my initial work that gets tabby running and using a llama-server binary instead

1. https://github.com/NixOS/nixpkgs/pull/334179
